crates/testing/src/byzantine/byzantine_behaviour.rs

```diff
@@ -358,3 +358,73 @@ impl<TYPES: NodeType> std::fmt::Debug for DishonestVoting<TYPES> {
             .finish_non_exhaustive()
     }
 }
+
+#[derive(Debug)]
+/// An `EventHandlerState` that will send a vote for a bad proposal
+pub struct DishonestVoter<TYPES: NodeType> {
+    /// Collect all votes the node sends
+    pub votes_sent: Vec<QuorumVote<TYPES>>,
+    /// Which vote to be dishonest at
+    pub dishonest_at_vote_numbers: HashSet<u64>,
+    /// How many votes from node
+    pub total_votes_from_node: u64,
+}
+
+#[async_trait]
+impl<TYPES: NodeType, I: NodeImplementation<TYPES> + std::fmt::Debug, V: Versions>
+    EventTransformerState<TYPES, I, V> for DishonestVoter<TYPES>
+{
+    async fn recv_handler(&mut self, event: &HotShotEvent<TYPES>) -> Vec<HotShotEvent<TYPES>> {
+        vec![event.clone()]
+    }
+
+    async fn send_handler(
+        &mut self,
+        event: &HotShotEvent<TYPES>,
+        public_key: &TYPES::SignatureKey,
+        private_key: &<TYPES::SignatureKey as SignatureKey>::PrivateKey,
+        upgrade_lock: &UpgradeLock<TYPES, V>,
+        _consensus: Arc<RwLock<Consensus<TYPES>>>,
+    ) -> Vec<HotShotEvent<TYPES>> {
+        match event {
+            HotShotEvent::QuorumProposalRecv(_proposal, _sender) => {
+                if self
+                    .dishonest_at_vote_numbers
+                    .contains(&self.total_votes_from_node)
```

Comment:
I think this should work, but it feels a bit fragile to me -- is it possible to use the view number of the proposal instead? I'm worried about e.g. a node sending out a double proposal throwing the test off if things change down the line

alternatively, for this specific scenario we could probably even share state with a cloned `RwLock` e.g. leader writes its bad proposals to an `RwLock`, follower always votes on proposals in that lock

Reply:
i like the idea of using an RwLock, view number also feels a bit fragile to me.

Reply:
i added a shared hashset between all Byzantine replicas that will store any view we send a dishonest proposal on. The Byzantine voting replicas will just check the current view in the set to see if they should send a qc vote and dont send a timeout vote.